hooks/open-telemetry/pkg/otel.go

```diff
@@ -0,0 +1,161 @@
+	AttributeTraceName = "github.com/open-feature/golang-sdk-contrib/hooks/open-telemetry/pkg"
+)
+
+type Hook struct {
+	spans        map[string]*mutexWrapper
+	ctx          context.Context
+	wg           *sync.WaitGroup
+	tracerClient tracerClientInterface
+}
+
+// mutex wrapper is used to prevent colliding keys from overwriting each other / closing a partially completed span
+type mutexWrapper struct {
+	mu *sync.Mutex
+	ss *storedSpan // ss has been set as a pointer so it can be set to nil during cleanup
+}
+type storedSpan struct {
+	cancel func()
+	span   trace.Span
+}
+
+// NewHook return a reference to a new instance of the OpenTelemetry Hook
+func NewHook() *Hook {
+	return &Hook{
+		tracerClient: &tracerClient{},
+	}
+}
+
+// Wait blocks until all spans have been closed
+func (h *Hook) Wait() {
+	h.wg.Wait()
+}
+
+// WithContext sets the parent context used when new spans are formed. When the parent context is closed so are all spans.
+// if no parent context is provided context.Background() is used internally
+func (h *Hook) WithContext(ctx context.Context) {
+	h.ctx = ctx
+}
+
+// Before creates the flag evaluations open-telemetry span and sets the FlagKey and ProviderName attributes
+func (h *Hook) Before(hookContext of.HookContext, hookHints of.HookHints) (*of.EvaluationContext, error) {
+	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
+	h.setup()
+	if _, ok := h.spans[key]; !ok {
+		h.spans[key] = &mutexWrapper{
+			mu: &sync.Mutex{},
+		}
+	}
+	h.spans[key].mu.Lock()
+	h.wg.Add(1)
+	ctx, span := h.tracerClient.tracer().Start(h.ctx, key)
+	ctx, cancel := context.WithCancel(ctx)
+	span.SetAttributes(
+		attribute.String(AttributeFlagKey, hookContext.FlagKey()),
+		attribute.String(AttributeProviderName, hookContext.ProviderMetadata().Name),
+	)
+	h.spans[key].ss = &storedSpan{
+		cancel: cancel,
+		span:   span,
+	}
+	// this goroutine cleans up the span, if the associated context is closed then the stored
+	// span data is removed and the resource is unlocked. This context close can come from either the cancel() method
+	// or from the closing of the parent context outside the scope of the hook
+	go func() {
+		<-ctx.Done()
+		h.spans[key].ss = nil
+		h.spans[key].mu.Unlock()
+		h.wg.Done()
+	}()
+	return nil, nil
+}
+
+// After sets the EvaluatedVariant and EvaluatedValue on the evaluation specific span
+func (h *Hook) After(hookContext of.HookContext, flagEvaluationDetails of.EvaluationDetails, hookHints of.HookHints) error {
+	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
+	mw, ok := h.spans[key]
+	if !ok {
+		return errors.New("no span stored for provided hook context")
+	}
+	attributes := []attribute.KeyValue{
+		attribute.String(AttributeEvaluatedVariant, flagEvaluationDetails.ResolutionDetail.Variant),
+	}
+	fmt.Println(flagEvaluationDetails)
+	switch flagEvaluationDetails.FlagType {
```

Comment:
Here's a link to the OTel spec proposal: 

https://github.com/open-telemetry/opentelemetry-specification/pull/2529/files

The `evaluated_value` should only be used if the `evaluated_variant` is unavailable. Also, the `evaluated_value` should be cast to a string.

Reply:
I have made these changes and added a unit test to ensure all returned values are of type string